### PR TITLE
Add a search tool to the web API

### PR DESCRIPTION
**Pull request by SamDepardieu**
# Description

This PR brings a search route to the API.

 - Define the route in api.py
 - Define the service in search.py
 - Change elastic/import_to_s3.py to also insert pdf_names and (randomly assigned for the time being) organisations
 - Add a persistent volume to the elasticsearch container

## Type of change
- [x] :sparkles: New feature

# How Has This Been Tested?

Mac OSx 10.14.4, Python3.6.4, Docker 18.09.5 with 2CPU/4Gb RAM

### CODE REVIEW

**Comment by SamDepardieu** [comment]
[Allow user to search a policy document by a REST API](https://trello.com/c/fUzpcFz1/670-allow-user-to-search-a-policy-document-by-a-rest-api)

File: policytool/web/search.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            else:
+                resp.body = json.dumps({
+                    'status': 'error',
+                    'message': response
+                })
+        else:
+            resp.body = json.dumps({
+                'status': 'error'
+            })
+            resp.status = falcon.HTTP_200
```
Comment (hblanks) [inline policytool/web/search.py]: Seems like this is an invalid request, so not 200?

File: policytool/web/api.py [hunk 2]
```diff
@@ -104,11 +112,31 @@ def on_get(self, req, resp):
...
+    logger.info('Trying to connect to {elastic_host}'.format(
+        elastic_host=os.environ['ELASTICSEARCH_HOST']
+    ))
+    es = Elasticsearch([{
+            'host': parsed_url.hostname,
+            'port': parsed_url.port
+        }],
+    )
+except Exception as e:
+    raise e
```
Comment (hblanks) [inline policytool/web/api.py]: Like Master Yoda would say, "There is no try" here, just do or do not :-)

File: policytool/web/api.py [hunk 3]
```diff
@@ -12,6 +12,12 @@
 import falcon
 import jinja2
 
+from urllib.parse import urlparse
+
+from elasticsearch import Elasticsearch
+
+from web import search
```
[the remark this answers is not in the document]
Comment (SamDepardieu) [reply policytool/web/api.py]: You can only use relative import inside a `from x import y` if I remember well, but I sure could use a `from . import search` 👍

File: policytool/web/search.py (hunk 1 above)
Comment (SamDepardieu) [reply policytool/web/search.py]: Whoopsy

File: policytool/web/search.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            return True, self.es.search(
+                index='datalabs-fulltexts',
+                body=json.dumps(es_body),
+                explain=True
+            )
+
+        except ConnectionError:
+            message = 'Could not join the elasticsearch server'
+            api.logger.error(message)
+            return False, message
```
[the remark this answers is not in the document]
Comment (SamDepardieu) [reply policytool/web/search.py]: If I raise it here, it makes the server crash when elasticsearch isn't ready yet to accept connections. The idea between the return was to fail without getting a 500

[the remark this answers is not in the document]
Comment (SamDepardieu) [reply policytool/web/search.py]: We should still set the status to a 50x (504 I think?) though

File: policytool/web/api.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hblanks) [reply policytool/web/api.py]: Either turns out to work -- search for "import ." in https://www.python.org/dev/peps/pep-0328/ for reference.

But, imports like this do only work when you're inside a "package." This is complicated but I can explain if you like some time.

File: policytool/web/search.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hblanks) [reply policytool/web/search.py]: Interesting! Does the server crash or just the handler for that HTTP request? I'd just taken exception handling for granted...

Docs say use falcon's HTTPError or some variant with the right status -- that sounds right to me. Cf. https://falcon.readthedocs.io/en/stable/api/errors.html . In that case the error logging is also very well placed; we'd rely on that for Sentry reporting.

I'd return a 503 b/c your dependency is not exactly an HTTP backend.

File: policytool/elastic/import_from_s3.py [hunk 5]
```diff
@@ -40,9 +41,15 @@ def write_to_es(es, line):
         line: a dict from a csv line. Should contain a file hash and the
               full text of a pdf
     """
+    # TODO: Use real pdf titles and orgs
+    # Select a random org for the time being
+    orgs = ['who', 'nice', 'msf']
+    org = random.choice(orgs)
```
Comment (nsorros) [inline policytool/elastic/import_from_s3.py]: what's going on here? :p

File: policytool/web/search.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            req_params: The request's parameters. Shoud include 'term' and at
+                        least a field (on of [text|title|organisation])
+
+        Returns:
+            True|False: The search success status
+            es.search()|str: A dict containing the result of the search if it
+                             succeeded or a string explaining why it failed
+        """
+        try:
+            fields = req_params.get('fields', [])
```
Comment (nsorros) [inline policytool/web/search.py]: maybe check that fields elements are valid

Comment (SamDepardieu) [reply policytool/web/search.py]: That's no issue if they are not, it'll just find no matches

File: policytool/elastic/import_from_s3.py (hunk 5 above)
Comment (SamDepardieu) [reply policytool/elastic/import_from_s3.py]: I don't have the real data in the dataset I'm unsing here, so I give it a random placeholder instead :D

File: policytool/elastic/import_from_s3.py [hunk 7]
```diff
@@ -5,6 +5,7 @@
 import tempfile
 import csv
 import json
+import random
```
Comment (hblanks) [inline policytool/elastic/import_from_s3.py]: @nsorros  - You asked a great question later in this file. To go one higher in your tech lead skills, just ask your reviewee

> what's going on here?

not at invocation time but at import.

**Pull request merged**